src/Control/Effect/Choose.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE DeriveGeneric, DeriveTraversable, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Choose
+( -- * Choose effect
+  Choose(..)
+, choose
+  -- * Choose carrier
+, runChoose
+, ChooseC(..)
+) where
+
+import Control.Applicative ((<|>), liftA2)
+import Control.Effect.Carrier
+import Control.Monad (join)
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Data.Bool (bool)
+import Data.Maybe (fromJust)
+import GHC.Generics (Generic1)
+
+data Choose m k
+  = Choose (Bool -> m k)
+  deriving (Functor, Generic1)
+
+instance HFunctor Choose
+instance Effect   Choose
+
+-- | Nondeterministically choose between two computations.
+choose :: (Carrier sig m, Member Choose sig) => m a -> m a -> m a
+choose a b = send (Choose (bool b a))
```

Comment:
I didn’t want to bring in a dependency on `semigroupoids` (and all of the transitive dependencies it entails), or this might have been an `Alt` instance (tho the n² instance problem would apply to that anyway).

NB: this name conflicts with #201, but I think we might want to generalize it to a `foldMapA` function instead anyway.